### PR TITLE
Fix #3926: Disallow implicit objects as parameter destructuring

### LINKED ISSUE #3926: Unexpected error during compilation
If compile the following code, the compiler crash unexpectedly :

```coffeescript
(toto : true) ->
    console.log "bug"

```

here is the partial call stack 

```
TypeError: Cannot read property 'length' of undefined
  at Param.exports.Param.Param.eachName (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:2123:30)
  at Param.exports.Param.Param.eachName (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:2126:16)
  at Code.exports.Code.Code.eachParamName (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:2047:28)
  at Code.exports.Code.Code.compileNode (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:2001:12)
  at Code.exports.Base.Base.compileToFragments (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:83:21)
  at Block.exports.Block.Block.compileNode (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:387:28)
  at Block.exports.Block.Block.compileWithDeclarations (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:480:19)
  at Block.exports.Block.Block.compileRoot (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:452:24)
  at Block.exports.Block.Block.compileToFragments (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:369:21)
  at /usr/local/lib/node_modules/coffee-script/lib/coffee-script/coffee-script.js:61:38
  at /usr/local/lib/node_modules/coffee-script/lib/coffee-script/coffee-script.js:34:19
```

If help is needed I might try to find some time and make a pull request.

Best regards and thank you for your help

Guillaume Leclerc

### PR DESCRIPTION
**Pull request by lydell**

### CODE REVIEW

File: src/nodes.coffee [hunk 1]
```diff
@@ -1482,6 +1482,9 @@ exports.Param = class Param extends Base
   constructor: (@name, @value, @splat) ->
     if (name = @name.unwrapAll().value) in STRICT_PROSCRIBED
       @name.error "parameter name \"#{name}\" is not allowed"
+    if @name instanceof Obj and @name.generated
+      token = @name.objects[0].operatorToken
```
Comment (vendethiel) [inline src/nodes.coffee]: is there no way that part of this could be null? (I suppose we don't have any degenerate case where we insert an empty object?)

Comment (lydell) [inline src/nodes.coffee]: An implicit object can’t be empty, can it? What would that look like in code?

**Comment by jashkenas** [comment]
Why is this disallowed?

**Comment by lydell** [comment]
Because #3926.

> I had no idea destructuring assignments didn't need brackets
> ...
> It is definitely confusing that (a: b) -> destructures, but a: b = obj does not.
> ...
> IMHO, I think that requiring brackets in this case would be a good thing.
> ...

etc. It is also not present anywhere in the docs, and isn’t used in CoffeeScript’s own source code.

```coffee
fn = (a, b = 1, c: d, e: f = 2, {g = 3}, h) -> ugly
```

**Comment by jashkenas** [comment]
Ah, I see.

**Pull request merged**